scripts/gulpfiles/build_tasks.js

```diff
@@ -418,6 +298,196 @@ function buildLangfiles(done) {
+  const rawOptions= JSON.parse(String(execSync(cccCommand)));
+
+  // rawOptions should now be of the form:
+  //
+  // {
+  //   chunk: [
+  //     'requires:258',
+  //     'all:10:requires',
+  //     'all1:11:requires',
+  //     'all2:11:requires',
+  //     /* ... remaining handful of chunks */
+  //   ],
+  //   js: [
+  //     '/Users/cpcallen/src/blockly/core/serialization/workspaces.js',
+  //     '/Users/cpcallen/src/blockly/core/serialization/variables.js',
+  //     /* ... remaining several hundred files */
+  //   ],
+  // }
+  //
+  // This is designed to be passed directly as-is as the options
+  // object to the Closure Compiler node API, but we want to replace
+  // the unhelpful entry-point based chunk names (let's call these
+  // "nicknames") with the ones from chunks.  Luckily they will be in
+  // the same order that the entry points were supplied in - i.e.,
+  // they correspond 1:1 with the entries in chunks.
+  const chunkByNickname = Object.create(null);
+  const chunkList = rawOptions.chunk.map((element, index) => {
+    const [nickname, numJsFiles, dependencyNicks] = element.split(':');
+    const chunk = chunks[index];
+    chunkByNickname[nickname] = chunk;
+    if (!dependencyNicks) {  // Chunk has no dependencies.
+      chunk.dependencies = [];
+      return `${chunk.name}:${numJsFiles}`;
+    }
+    chunk.dependencies =
+        dependencyNicks.split(',').map(nick => chunkByNickname[nick]);
+    const dependencyNames =
+        chunk.dependencies.map(dependency => dependency.name).join(',');
+    return `${chunk.name}:${numJsFiles}:${dependencyNames}`;
+  });
+
+  // Generate a chunk wrapper for each chunk.
+  for (const chunk of chunks) {
+    chunk.wrapper = chunkWrapper(chunk);
+  }
+
+  const chunkWrappers = chunks.map(chunk => `${chunk.name}:${chunk.wrapper}`);
+  return {chunk: chunkList, js: rawOptions.js, chunk_wrapper: chunkWrappers};
+}
+
+/** 
+ * RegExp that globally matches path.sep (i.e., "/" or "\").
+ */
+const pathSepRegExp = new RegExp(path.sep.replace(/\\/, '\\\\'), "g");
+
+/** 
+ * Modify the supplied gulp.rename path object to relax @package
+ * restrictions in core/.
+ *
+ * Background: subdirectories of core/ are used to group similar files
+ * together but are not intended to limit access to names
+ * marked @package; instead, that annotation is intended to mean only
+ * that the annotated name not part of the public API.
+ *
+ * To make @package behave less strictly in core/, this function can
+ * be used to as a gulp.rename filter, modifying the path object to
+ * flatten all files in core/** so that they're in the same directory,
+ * while ensuring that files with the same base name don't conflict.
+ *
+ * @param {{dirname: string, basename: string, extname: string}}
+ *     pathObject The path argument supplied by gulp.rename to its
+ *     callback.  Modified in place.
+ */
+function flattenCorePaths(pathObject) {
```

Comment:
Some files are multiple directories down (e.g. renderers). Does this run multiple times?

Reply:
After reading the rest: no, it doesn't run multiple times. I think it needs to loop until it runs out of slashes.

Reply:
I am pretty sure the code is correct.  After the call to `.split(path.sep)`, there will be no slashes in the `dirs` array.  `basename` is then reassembled using `.join`.

To check, I added an error to `core/renderers/zelos/measurables/inputs.js`, and the resulting error message was for `core/renderers-slash-zelos-slash-measurables-slash-inputs.js`.

(Note that we do not want to get rid of the slash after `core`, because we only want to expand the scope of `@package` within `core`, and not between `core` and any other directory.)